ccl/changefeedccl: TestChangefeedNemeses failed

ccl/changefeedccl.TestChangefeedNemeses [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsGcpLinuxX8664BigVm_CclUnitTests/11186536?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsGcpLinuxX8664BigVm_CclUnitTests/11186536?buildTab=artifacts#/) on master @ [d87a04bf528cf3d8f242ceb539451cd14aa7a3cf](https://github.com/cockroachdb/cockroach/commits/d87a04bf528cf3d8f242ceb539451cd14aa7a3cf):

Fatal error:

```
panic: test timed out after 14m55s

```
Stack: 

```
goroutine 532466 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x8e
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x32
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestChangefeedNemeses
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/da00bb69111b8c7016260e5710f1cc2b/logTestChangefeedNemeses332822107
    test_log_scope.go:81: use -show-logs to present logs inline
=== RUN   TestChangefeedNemeses/cloudstorage
    helpers_test.go:858: making server as system tenant
    helpers_test.go:936: making cloudstorage feed factory
    nemeses_test.go:32: random seed: -4423574052186727118
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedNemeses.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-30333

--- a/pkg/ccl/changefeedccl/cdctest/nemeses.go
+++ b/pkg/ccl/changefeedccl/cdctest/nemeses.go
@@ -30,7 +30,7 @@ import (
 // real output of a changefeed. The output rows and resolved timestamps of the
 // tested feed are fed into them to check for anomalies.
 func RunNemesis(
-	f TestFeedFactory, db *gosql.DB, isSinkless bool, rng *rand.Rand,
+	f TestFeedFactory, db *gosql.DB, isSinkless bool, withLegacySchemaChanger bool, rng *rand.Rand,
 ) (Validator, error) {
 	// possible additional nemeses:
 	// - schema changes
@@ -54,9 +54,10 @@ func RunNemesis(
 		eventPauseCount = 0
 	}
 	ns := &nemeses{
-		maxTestColumnCount: 10,
-		rowCount:           4,
-		db:                 db,
+		withLegacySchemaChanger: withLegacySchemaChanger,
+		maxTestColumnCount:      10,
+		rowCount:                4,
+		db:                      db,
 		// eventMix does not have to add to 100
 		eventMix: map[fsm.Event]int{
 			// We don't want `eventFinished` to ever be returned by `nextEvent` so we set
@@ -262,6 +263,8 @@ type addColumnPayload struct {
 }
 
 type nemeses struct {
+	withLegacySchemaChanger bool
+
 	rowCount           int
 	maxTestColumnCount int
 	eventMix           map[fsm.Event]int
@@ -692,9 +695,14 @@ func addColumn(a fsm.Args) error {
 	if err := ns.db.QueryRow(`SELECT count(*) FROM foo`).Scan(&rows); err != nil {
 		return err
 	}
-	// We expect one table scan that corresponds to the schema change backfill, and one
-	// scan that corresponds to the changefeed level backfill.
-	ns.availableRows += 2 * rows
+	if ns.withLegacySchemaChanger {
+		// We expect one table scan that corresponds to the schema change backfill, and one
+		// scan that corresponds to the changefeed level backfill.
+		ns.availableRows += 2 * rows
+	} else {
+		// We expect to see a backfill
+		ns.availableRows += rows
+	}
 	return nil
 }
 
@@ -715,9 +723,14 @@ func removeColumn(a fsm.Args) error {
 	if err := ns.db.QueryRow(`SELECT count(*) FROM foo`).Scan(&rows); err != nil {
 		return err
 	}
-	// We expect one table scan that corresponds to the schema change backfill, and one
-	// scan that corresponds to the changefeed level backfill.
-	ns.availableRows += 2 * rows
+	if ns.withLegacySchemaChanger {
+		// We expect one table scan that corresponds to the schema change backfill, and one
+		// scan that corresponds to the changefeed level backfill.
+		ns.availableRows += 2 * rows
+	} else {
+		// We expect to see a backfill
+		ns.availableRows += rows
+	}
 	return nil
 }
 

--- a/pkg/ccl/changefeedccl/nemeses_test.go
+++ b/pkg/ccl/changefeedccl/nemeses_test.go
@@ -32,11 +32,11 @@ func TestChangefeedNemeses(t *testing.T) {
 		t.Logf("random seed: %d", seed)
 
 		sqlDB := sqlutils.MakeSQLRunner(s.DB)
-		_ = maybeDisableDeclarativeSchemaChangesForTest(t, sqlDB, rng)
+		withLegacySchemaChanger := maybeDisableDeclarativeSchemaChangesForTest(t, sqlDB, rng)
 		// TODO(dan): Ugly hack to disable `eventPause` in sinkless feeds. See comment in
 		// `RunNemesis` for details.
 		isSinkless := strings.Contains(t.Name(), "sinkless")
-		v, err := cdctest.RunNemesis(f, s.DB, isSinkless, rng)
+		v, err := cdctest.RunNemesis(f, s.DB, isSinkless, withLegacySchemaChanger, rng)
 		if err != nil {
 			t.Fatalf("%+v", err)
 		}

